pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java

```diff
@@ -0,0 +1,183 @@
+ * under the License.
+ */
+
+package org.apache.pinot.controller.recommender.rules.impl;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.pinot.controller.recommender.exceptions.InvalidInputException;
+import org.apache.pinot.controller.recommender.io.ConfigManager;
+import org.apache.pinot.controller.recommender.io.InputManager;
+import org.apache.pinot.controller.recommender.realtime.provisioning.MemoryEstimator;
+import org.apache.pinot.controller.recommender.rules.AbstractRule;
+import org.apache.pinot.controller.recommender.rules.io.configs.IndexConfig;
+import org.apache.pinot.controller.recommender.rules.io.params.RealtimeProvisioningRuleParams;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.DataSizeUtils;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+
+
+/**
+ * This rule gives some recommendations useful for provisioning real time tables. Specifically it provides some
+ * recommendations for optimal segments size, total memory used per host, and consuming memory used per host based
+ * on the provided characteristics of the data.
+ */
+public class RealtimeProvisioningRule extends AbstractRule {
+  public static String OPTIMAL_SEGMENT_SIZE = "Optimal Segment Size";
+  public static String CONSUMING_MEMORY_PER_HOST = "Consuming Memory per Host";
+  public static String TOTAL_MEMORY_USED_PER_HOST = "Total Memory Used per Host";
+
+  private final RealtimeProvisioningRuleParams _params;
+
+  public RealtimeProvisioningRule(InputManager input, ConfigManager output) {
+    super(input, output);
+    _params = input.getRealtimeProvisioningRuleParams();
+  }
+
+  @Override
+  public void run()
+      throws InvalidInputException {
+
+    if (_params == null) {
+      // no realtime provisioning params provided; skip
+      return;
+    }
+
+    // prepare input to memory estimator
+    TableConfig tableConfig = createTableConfig(_output.getIndexConfig(), _input.getSchema());
+    long maxUsableHostMemoryByte = DataSizeUtils.toBytes(_params.getMaxUsableHostMemory());
+    int totalConsumingPartitions = _params.getNumPartitions() * _params.getNumReplicas();
+    int ingestionRate = (int) _input.getNumMessagesPerSecInKafkaTopic() / _params.getNumPartitions();
+    int[] numHosts = _params.getNumHosts();
+    int[] numHours = _params.getNumHours();
+
+    // run memory estimator
+    MemoryEstimator memoryEstimator =
+        new MemoryEstimator(tableConfig, _input.getSchema(), _input.getSchemaWithMetadata(),
+            (int) _input.getNumRecordsPerPush(), ingestionRate, maxUsableHostMemoryByte,
```

Comment:
Please consider adding a TODO here to indicate that we may not want to use numRecordsPerPush as the numRows for the completed segment we are going to generate. A more fine-grained number is needed which we need to figure out how to capture.